Add initial extension

rly opened the pull request:
- [x] Create new extension from latest [ndx-template](https://github.com/nwb-extensions/ndx-template)
- [x] Add new spec types in `create_extension_spec.py`
- [x] Generate specs YAML
- [x] Test auto-generation of classes from schema using `get_class`
- [x] Add unit tests
- [x] Add integration tests
- [x] Expand example
- [x] Test example code
- [x] ~~Resolve https://github.com/hdmf-dev/hdmf/issues/1099 before adding `ProbeInsertion.insertion_position_in_mm` and `ProbeInsertion.insertion_angle_in_deg`~~
- [x] Resolve relevant TODOs in `pyproject.yaml`
- [x] Resolve relevant TODOs in `create_extension_spec.py`
- [x] Update mermaid UML
- [x] Improve docstrings
- [x] Adapt custom API code from https://github.com/SpikeInterface/ndx-probeinterface/tree/main/src/pynwb

rly committed:
Add initial extension without tests or custom api

CodyCBakerPhD committed:
Merge branch 'main' into initial-extension

CodyCBakerPhD committed:
Update README.md

rly committed:
Add tests, small updates to spec

rly committed:
Add "in_mm", add example usage, update tests

CodyCBakerPhD commented:
Wow, this is really coming together! Just let me know when ready for an initial review - I'm OK with quick merge cycles and subsequent fixes/improvements as well

rly commented:
Good point. This PR doesn't need to be perfect and I think it is good enough for initial review. We can tackle the remaining TODO items in separate PRs.

rly commented:
Actually before review, I'm going to make a few small organizational changes.

CodyCBakerPhD commented:
> Actually before review, I'm going to make a few small organizational changes.

Sure, what kind of organization changes were you thinking?

h-mayorquin commented:
Let me know when I should take a look into this.

rly committed:
Add comments

rly commented:
OK. Sorry for my delay; I admit, I think I sent that last message on a plane and no longer remember what else I wanted to change. So this is ready for review.

CodyCBakerPhD commented on README.md:
```diff
@@ -1,6 +1,71 @@
+
+## Usage
+
+(TODO test these examples)
```
Were tests added for these?

rly replied:
I just added some tests now

CodyCBakerPhD commented on notebooks/example.ipynb:
Looks like this file can be removed

CodyCBakerPhD commented:
Just a couple minor points

After that happy to merge this and raise issues as follow-ups if bugs are encountered or changes requested

rly committed:
Delete notebooks/example.ipynb

rly commented on src/pynwb/ndx_extracellular_channels/widgets/tetrode_series_widget.py:
That's a good idea.

rly committed:
Remove references to tetrodeseries

rly commented on spec/ndx-extracellular-channels.extensions.yaml:
```diff
@@ -0,0 +1,293 @@
+      - 2
+    - - null
+      - 3
+    doc: The planar polygon that outlines the probe contour.
```
```suggestion
    doc: The coordinates of the nodes of the polygon that describe the shape (contour) of the probe,
      e.g., [(-20, -30), (20, -110), (60, -30), (60, 190), (-20, 190)].
```

rly commented on spec/ndx-extracellular-channels.extensions.yaml:
```diff
@@ -0,0 +1,293 @@
+  - name: planar_contour
+    dtype: float
+    dims:
+    - - num_points
+      - x
```
It represents coordinates of the nodes of the polygon that describe the shape (contour) of the probe. I expanded the docstring to say that

CodyCBakerPhD replied:
The example in the docstring is for 2D

Can you provide me an example of a 1D polygon?

rly replied:
Oh, I understand now. Sorry about that. I will check with Alessio since this was copied from ndx-probeinterface

alejoe91 replied:
Yeah I think this must be at least 2D

rly committed:
Update spec/ndx-extracellular-channels.extensions.yaml

rly committed:
Clarify planar_contour

rly committed:
Minor refactor, support probeinterface IO

rly committed:
Fix lint

rly committed:
Add probeinterface to dev env

rly committed:
Fix probeinterface import

rly commented:
@CodyCBakerPhD I adapted the original `from_probeinterface` and `to_probeinterface` methods and made some minor changes to the extension:
- removed `ProbeModel.model_name` - it is redundant with `ProbeMode.name`
- made `planar_contour` an attribute instead of a dataset
- removed the ability to set the unit on `planar_contour`, so it is now `planar_contour_in_um`
- made slight changes to the names of a couple fields

In retrospect, I should have requested approval, merged, and added the probeinterface tests and minor refactor in a separate PR. Sorry about that. I am happy to revert these changes and make a separate PR with the changes. Or you could review the changes here. What would you prefer?

CodyCBakerPhD commented:
> removed ProbeModel.model_name - it is redundant with ProbeMode.name

I would disagree; the `name` of the object is just a local reference to that neurodata object and must be unique across the file. I get that we would not usually have more than one abstraction of a probe model per model identifier, but the name field in general is so loose in NWB that I don't think we should make it the most important identifier of that piece of information

We are also looking to use `model` (not `model_name`) prolifically in the microscopy device extension

CodyCBakerPhD commented:
I'll just give the whole thing a re-review

With the `from` and `to` methods for PI, does @h-mayorquin have anything to do on your side with this?

rly commented:
> > removed ProbeModel.model_name - it is redundant with ProbeMode.name
> 
> I would disagree; the `name` of the object is just a local reference to that neurodata object and must be unique across the file. I get that we would not usually have more than one abstraction of a probe model per model identifier, but the name field in general is so loose in NWB that I don't think we should make it the most important identifier of that piece of information
> 
> We are also looking to use `model` (not `model_name`) prolifically in the microscopy device extension

That makes sense. Do you suggest we use `model` instead of `model_name`?

rly commented:
> With the `from` and `to` methods for PI, does @h-mayorquin have anything to do on your side with this?

No. These were adapted from ndx-probeinterface: https://github.com/SpikeInterface/ndx-probeinterface/blob/main/src/pynwb/ndx_probeinterface/io.py

CodyCBakerPhD commented:
> That makes sense. Do you suggest we use model instead of model_name?

For consistency, I think they should all agree - as to whether or not `model_name` or `model` or any other key phrase anyone can think of is preferred, I'm not sure. Maybe bring up at next LBNL meting?

rly commented:
Sounds good. Let's go with `model` for now, just like `manufacturer`, and we can discuss at the next sync.

rly committed:
Add back ProbeModel.model, remove 1D contour case

rly committed:
Update ruff config to latest recommended usage

CodyCBakerPhD commented on spec/ndx-extracellular-channels.extensions.yaml:
```diff
@@ -0,0 +1,287 @@
+    dtype: float
+    doc: Height of a rectangular contact, in micrometers.
+    quantity: '?'
+  - name: device_channel_index_pi
```
I read this as 'device channel index $\pi$'

Is that the intention?

If 'pi' is abbreviated for ProbeInterface for some reason, I would not abbreviate and just give the full context

rly replied:
From what I can tell, this is the index of the channel connected to the contact on the device. I changed the name to just `device_channel`.

alejoe91 replied:
Initially we used the `pi` suffix because everything that ends with `index` is handled differently

CodyCBakerPhD commented on spec/ndx-extracellular-channels.extensions.yaml:
```diff
@@ -0,0 +1,287 @@
+  - name: device_channel_index_pi
+    neurodata_type_inc: VectorData
+    dtype: int
+    doc: Index of the channel connected to the contact.
```
I would also emphasize why the extra bit is needed after `_index` and what  it means

CodyCBakerPhD commented on spec/ndx-extracellular-channels.extensions.yaml:
```diff
@@ -0,0 +1,287 @@
+    doc: The filter used on the raw (wideband) voltage data from this contact, including
+      the filter name and frequency cutoffs, e.g., 'High-pass filter at 300 Hz.'
+    quantity: '?'
+  - name: estimated_position_in_mm
```
It looks as if you may have changed various things in the schema since last we met

But the mermaid has not been updated; can you bring the two in sync?

rly replied:
Done

CodyCBakerPhD commented on spec/ndx-extracellular-channels.extensions.yaml:
```diff
@@ -0,0 +1,287 @@
+    dtype: text
+    doc: The brain area of the estimated contact position, e.g., 'CA1'.
+    quantity: '?'
+  - name: actual_position_in_mm
```
The word 'actual' here sounds a little odd. Maybe grants too much confidence? Even histology registration is still a form of estimation, just a higher resolution one

Perhaps 'verified'? or 'confirmed'?

As the opposite to 'estimated' (which I think is fine above), Google says 'confirmed' might be the best

rly replied:
I changed "actual" to "confirmed" throughout.

CodyCBakerPhD commented on src/pynwb/tests/test_example_usage_probeinterface.py:
```diff
@@ -0,0 +1,91 @@
+    io.write(nwbfile)
+
+# read the file and check the content
+with pynwb.NWBHDF5IO("test_probeinterface.nwb", "r", load_namespaces=True) as io:
```
Well, I suppose one think @h-mayorquin could still do here is add a mirror of these tests over on ProbeInterface side to ensure that any changes to the package over there don't break these helper functions in the extension

CodyCBakerPhD replied:
Also, is this really a test? It's in the 'tests' directory and starts with 'test_' but I don't see any assertions beyond 'does this succeed' (not 'does it succeed with a certain output')

Also would it run with pytest?

rly replied:
You're right - this is not a test and does not run with pytest (I had thought it would). I think if I wrapped it in a `test_x` function, it would run. I added asserts. I still need to add a test for the `to_probeinterface` function.

rly committed:
Change name of device_channel

rly committed:
Rename actual_x to confirmed_x

rly committed:
Restructure position fields

rly committed:
Rearrange spec, update readme

rly committed:
Update probeinterface test

rly committed:
Apply black

rly commented:
I changed `insertion_position_in_mm` from a compound data type of [float, float, float] for [ap, ml, dv] to three values `insertion_position_ap_in_mm`, `insertion_position_ml_in_mm`, and `insertion_position_dv_in_mm` because
1) compound dtype attributes are not supported in HDMF yet
2) matnwb does not handle compound dtypes well https://github.com/NeurodataWithoutBorders/matnwb/issues/572
3) compound dtype values/arrays are uncommon and confuse users on read and write. we could extract these at the API level, but I would prefer to keep the API consistent with the storage because it is less confusing to users and maintainers, and matnwb generates its API based on the storage anyway
4) in pynwb, you can specify the values for a compound dtype like `insertion_position_in_mm` as a tuple of three values without labels on them. this results in non-self-documenting code. which value is which? we could require users provide a numpy struct array with the keys "ap", "dv", and "ml" but that would be a pain
5) initially i had thought a user should supply all three values if they have one, but that may be too strict of a requirement in general. may be better to leave it up to the user.

rly committed:
Update docstring

rly committed:
Update comments

rly committed:
Update doc

ehennestad commented:
> matnwb does not handle compound dtypes well [Bug]: Pixel_mask doesn't seem to be properly constructed.  NeurodataWithoutBorders/matnwb#572

The problem in that issue was not really that matnwb did not handle compound data types, instead a lack of documentation for how to create pixel masks, and specifically a misunderstanding of how to specify the `pixel_mask_index` vector

CodyCBakerPhD commented on src/pynwb/tests/test_example_usage_all.py:
Same question for this file as the other one; doesn't look like it's being run by the testing suite?

rly replied:
Good catch. Fixed. Thanks

rly committed:
Add asserts to test all

rly committed:
Fix probeinterface converter shape keys

rly committed:
Remove print

rly committed:
Apply black, ruff, isort

rly commented:
> > matnwb does not handle compound dtypes well [Bug]: Pixel_mask doesn't seem to be properly constructed.  [NeurodataWithoutBorders/matnwb#572](https://github.com/NeurodataWithoutBorders/matnwb/issues/572)
> 
> The problem in that issue was not really that matnwb did not handle compound data types, instead a lack of documentation for how to create pixel masks, and specifically a misunderstanding of how to specify the `pixel_mask_index` vector

@ehennestad thanks for the clarification!

CodyCBakerPhD commented:
@rly Are all the 'TODO's in the PR message complete?

h-mayorquin commented:
I will do a review next week.

rly commented:
Remaining TODOs for discussion.

- `ContactsTable.relative_position_in_mm.reference` ("Reference point for the relative position coordinates and information about the coordinate system used, e.g., which direction is positive in the x direction (first element), which direction is positive in the y direction (second element), etc.") is not required, but I wonder if it should be for maximal interpretability, especially if a tool builder wants to use these coordinates with the insertion position, angle, and depth of the probe to compute the stereotactic coordinates of each contact. 

	ProbeInterface does not provide this information, so requiring this would break automatic conversion or require a placeholder. Given that and given the rarity of this field's use, I suggest we keep it optional.

- `ProbeModel.planar_contour_in_um` - what do you think about naming this `ProbeModel.contour_in_um`? "Planar" suggests 2D, but this can be 3D. This comes from `probe_planar_contour` in https://probeinterface.readthedocs.io/en/main/format_spec.html . 

- Should `ChannelsTable.contact` (the reference to rows of the `ContactsTable`) be optional? In other words, should a `ContactsTable` be required to create an `ExtracellularSeries`. I think probably. We currently require `ElectrodeGroup` and `Device`. I'm just thinking about the number of lines of code this will add to the tutorials...

rly committed:
Make contact required, minor other fixes

rly committed:
Improve eseries api for unit, fix value to microvolts

rly committed:
Make plane_axes realistic

rly committed:
Fix

rly committed:
Fix mermaid typo

rly committed:
Fix mermaid typo

CodyCBakerPhD commented:
> ContactsTable.relative_position_in_mm.reference ("Reference point for the relative position coordinates and information about the coordinate system used, e.g., which direction is positive in the x direction (first element), which direction is positive in the y direction (second element), etc.") is not required, but I wonder if it should be for maximal interpretability, especially if a tool builder wants to use these coordinates with the insertion position, angle, and depth of the probe to compute the stereotactic coordinates of each contact.
> 
> ProbeInterface does not provide this information, so requiring this would break automatic conversion or require a placeholder. Given that and given the rarity of this field's use, I suggest we keep it optional.

Could be required keyword argument of the conversion process (so still automated, just not with zero extra `kwargs`)

Otherwise it's as you say, the classic argument against making something required because some will fill with placeholders or leave empty `""`

So usual approach is to have optional and make NWB Inspector check to strongly suggest its inclusion


> ProbeModel.planar_contour_in_um - what do you think about naming this ProbeModel.contour_in_um? "Planar" suggests 2D, but this can be 3D. This comes from probe_planar_contour in https://probeinterface.readthedocs.io/en/main/format_spec.html .

I lean towards whichever is closest to ProbeInterface

> Should ChannelsTable.contact (the reference to rows of the ContactsTable) be optional? In other words, should a ContactsTable be required to create an ExtracellularSeries. I think probably. We currently require ElectrodeGroup and Device. I'm just thinking about the number of lines of code this will add to the tutorials...

I don't think so, it should be required even if it's an identity mapping

If that seems inconvenient, we can make an identity mapping the default via the API writing tools so that it's zero extra work for anyone using a classic probe setup

CodyCBakerPhD commented:
This looks good to me - will wait on @h-mayorquin for final review

h-mayorquin commented on src/pynwb/tests/test_example_usage_all.py:
```diff
@@ -0,0 +1,242 @@
+        position_confirmation_method="Histology",
+        probe=probe,
+        probe_insertion=pi,
+        target_tables={
+            "contact": probe.probe_model.contacts_table,
+            "reference_contact": probe.probe_model.contacts_table,
+        },
```
Is this a validation issue?

rly replied:
Good point. Since `probe`, `probe.probe_model`, and `probe_model.contacts_table` are all required, we can do this. I'll update the PR.

rly commented on src/pynwb/ndx_extracellular_channels/__init__.py:
```diff
@@ -0,0 +1,59 @@
+)
+
+# Remove these functions from the package
+del load_namespaces, get_class, init_dv
```
Good question. I don't think it hurts to have it to prevent someone from doing `from ndx_extracellular_channels import load_namespaces`. I think I put these into the template based on the `del` statements in https://github.com/numpy/numpy/blob/main/numpy/__init__.py and https://github.com/scipy/scipy/blob/main/scipy/__init__.py

rly replied:
But yes, it's not really necessary. Just being extra tidy.

h-mayorquin replied:
Thanks. I was just curios. I do remember it was in the base template.

h-mayorquin commented on src/spec/create_extension_spec.py:
```diff
@@ -0,0 +1,531 @@
+
+    probe = NWBGroupSpec(
+        neurodata_type_def="Probe",
+        neurodata_type_inc="Device",
```
There was an interesting debate in the in-person discussion about making a distinction between Devices that would be something similar to a class (the model that you can look at a catalogue of a manufacturer) and instances of them that would be the actual physical sensors on the experimental arrangment. 

It made sense to me that the ProbeModel would be a Device but it was unclear what type the (instance) Probe would be.

This probably is out of the scope here and I don't see any downside about defining both as subclasses of Device but I was still curios to hear your thoughts about it. Do we gain anything by making such distinction?

rly replied:
In NWB currently, the two concepts for Device are mixed together. If someone has two Neuropixels probes, they will create two Device objects, one for each probe, with different names. This Device has a `manufacturer` that are really a property of the model, not the instance. 

I think by making such a distinction, we reduce redundancy and improve interoperability with libraries like probeinterface and any catalogs of microscopes or other devices. It would be useful to have a `DeviceModel` type in core NWB. That's a bigger change for NWB that I think is beyond the scope of this extension.

h-mayorquin replied:
Cool. This makes sense to me. 
> It would be useful to have a DeviceModel type in core NWB. That's a bigger change for NWB that I think is beyond the scope of this extension.

Agree this is out of scope. Thanks for sharing your thoughts. I also think this distinction  makes sense in Probeinterface. I will keep thinking about it.

h-mayorquin commented on src/spec/create_extension_spec.py:
```diff
@@ -0,0 +1,531 @@
+        attributes=[
+            # inherits name, description, manufacturer from Device
+            NWBAttributeSpec(name="ndim", doc="dimension of the probe", dtype="int", default_value=2),
+            NWBAttributeSpec(
```
Could we modify ProbeModel so name is just the model by default unless the user specify it?

rly replied:
Good idea. I'll make that change.

h-mayorquin commented on src/spec/create_extension_spec.py:
```diff
@@ -0,0 +1,531 @@
+                dtype="text",
+            ),
+            NWBAttributeSpec(
+                name="planar_contour_in_um",  # TODO should this just be "contour"?
```
Yeah, I think planar also is kind of redundant. I think boundary_countour is better.

For reference this is the way that open cv uses contours is the just the list of vertices that define a boundary but I think matplotlib has another meaning. 

Not sure here, maybe @alejoe91 can sheed some ligth on why they decide to emphasize the planar part.

h-mayorquin commented on src/pynwb/tests/test_example_usage_all.py:
```diff
@@ -0,0 +1,242 @@
+
+    es = ExtracellularSeries(
+        name="ExtracellularSeries",
+        data=[0.0, 1.0, 2.0],
```
I am confused here. If I read well this represents data with one channel. However, you adding as a dataset a region that points to three channels. Is that right? Should it not throw an error because the region is pointing to does not match the number of channels that the ExtracellularSeries represents?

rly replied:
Good catch! Yes, you are right.

h-mayorquin commented on src/spec/create_extension_spec.py:
```diff
@@ -0,0 +1,531 @@
+                    "in micrometers. The first and last points are connected to close the polygon. "
+                    "e.g., [(-20., -30.), (20., -110.), (60., -30.), (60., 190.), (-20., 190.)]."
+                    "See 'probe_planar_contour' in "
+                    "https://probeinterface.readthedocs.io/en/main/format_spec.html for more details."
```
This link is not very clarifying but I think this is on probeinterface. We should probably a better explanation there.

rly replied:
I agree. I included the link to indicate that we intend this field to map to that field. I think the description here is pretty good, but please feel free to improve it here (or there).

h-mayorquin replied:
The description is here is good. It is on not good on probeinterface side.

h-mayorquin commented on src/spec/create_extension_spec.py:
```diff
@@ -0,0 +1,531 @@
+                dtype="text",
+                quantity="?",
+            ),
+            NWBDatasetSpec(
```
I think there is another knot here.

If I remember well one of the purposes of separating Probe from ProbeModel is so the latter represents the catalogue version of the Probe which is an instance in the experimental setup.

The contacts table falls firmly on the catalogue because we are assuming that Probes come with a consisten geometry from the catalogue.

However, the mapping from the contacts to the channels here is clearly a property of the experimental setup so I don't think it makes a lot of conceptual sense with the previous intent.

That said, I understand that this is for the round-trip to work and we practicality beats purity so I think is fine.

To be honest, I don't really like the way that the wiring is managed in the interface between probe interface and spikeinterface. I feel that the mapping should be done from the recording to the probe and not the other way around. Plus, I think that it should be done from the channel ids (in spikeinterface) to the contacts ids (in probeinterface) instead of using a mapping from indices to indices which is an implementation detail of the data structures in both data models. 

I have not had the time to pitch these concents to Sam and Alessio but I think I should do this soon.

rly replied:
The headstage (or data acquisition system) has a particular way it maps channels (0 to N-1) to contacts. This mapping is a property of the headstage. We don't really represent the headstage in NWB except with properties like filters in the `ChannelsTable` (or ElectrodesTable - see #3).

I think this column was introduced by @alejoe91 and @khl02007 in the initial iteration of ndx-probeinterface. I agree that it doesn't really belong on the `ProbeModel` and is more appropriate on `Probe` or `ChannelsTable`. Should we remove this for now? We can always add it back later once we are happy with it.

h-mayorquin replied:
No, I am not confused about what it is. I understand that there is a mapping from the DAQ/Controller to the probe (that sometimes passess through a headstage). We need that mapping and in probeinterface case is done through that field.

But I think that in our conceptual model it does not make sense to put the mapping there. So yes, I think we should remove it. We already have `contact` as a column in the channels table that is taking care of that in what I think is a conceptually cleaner way.

rly replied:
> We already have `contact` as a column in the channels table that is taking care of that in what I think is a conceptually cleaner way.

Yeah, I think that is the better way at least in our model. 

OK will do.

h-mayorquin commented on src/spec/create_extension_spec.py:
```diff
@@ -0,0 +1,531 @@
+                required=False,
+            ),
+            NWBAttributeSpec(
+                name="reference_mode",
```
Given that we are thinking on adding `position_reference` could we change this to `electrical_reference_mode` or something of the sort? What do you guys think? I think is more clear even if `position_refrence` did not exist.

rly replied:
I like `electrical_reference_mode`. I will make that change

h-mayorquin replied:
Great

h-mayorquin commented on src/spec/create_extension_spec.py:
```diff
@@ -0,0 +1,531 @@
+                neurodata_type_inc="DynamicTableRegion",
+                doc="The row in a ContactsTable that represents the contact used as a channel.",
+            ),
+            NWBDatasetSpec(
```
How does this interact with the current `reference_mode`? Are the suppoed to be mutually exclusive? As in, you can specify which channel is used as reference and/or `reference_mode` accounts for the cases when there is a global reference?

rly replied:
Thanks for bringing this up. Yeah, the documentation for `reference_mode` here is unclear. 

`reference_mode` currently does two things: 
1. capture the point of reference for getting the voltage. This is commonly referred to as the ground (e.g., skull screw, grounding wire)
2. capture any special referencing done, e.g., "common average referencing" which would mean the data across all channels is averaged and subtracted from the current channel before being recorded. Some data acquisition systems can do this online and the referenced data is what is stored.

In reviewing this, I think we should have two optional fields here `ground` and `electrical_reference_description` for each of the above. I'll make that change.

If a user supplies a `reference_channel` for each channel, then `electrical_reference_description` is not needed, but could be useful as a freeform description to explain why the references were set the way they were set.

This schema does not cover the rare case where channels 1-16 have channel 17 as reference and channels 18-32 have a common average reference using only those channels, because we cannot specify None as a reference channel for only some channels...

https://plexon.com/blog-post/point-of-reference/
https://open-ephys.github.io/gui-docs/User-Manual/Plugins/Neuropixels-PXI.html#reference-selection
https://github.com/cortex-lab/neuropixels/wiki/Referencing-and-Grounding
https://www.youtube.com/watch?v=wKx_B5PR3YU&t=489s

h-mayorquin replied:
Yeah, just checked the new version. Looks great. Also, thanks for those links.

Out of curiosity:
Concerning the limitation that you mentioned, is that a limitation because the dtype of `DynamicTableRegion` is an int and we don't have nullable ints, right? 

On the other hand, given that this is supposed to represent indices, could we just use a negative number to indicate that?

rly replied:
> a limitation because the dtype of DynamicTableRegion is an int and we don't have nullable ints, right?

Right

> On the other hand, given that this is supposed to represent indices, could we just use a negative number to indicate that?

Yes, we could. If we want to do that, then we should codify that in https://github.com/hdmf-dev/hdmf-common-schema/blob/main/common/table.yaml#L71 and guard for that in the HDMF API. I'm pretty sure a -1 row index would result in an error right now.

I think we do not need to do this until a user requests it. I think it is a pretty rare use case.

h-mayorquin replied:
OK, thanks for explaining this to me.

h-mayorquin commented on src/pynwb/tests/test_example_usage_all.py:
```diff
@@ -0,0 +1,242 @@
+    # all of the keyword arguments in add_row are optional
+    channels_table.add_row(
+        contact=0,
+        reference_contact=2,
```
This is the question about the spec and how it works I think. 
I am confused about how we are adding integers here. The spec says that `contact` inherits from `DynamicTableRegion` and I was expecting that I would have to pass a `DyanmicTableRegion` object here. 

How does it work? Is this a synatactic suggar that creates a region with one entry under the hood? Would it work if I pass a Dynamic Region? I can test the later and I will do now but I wanted to ask in case I got stuck : P

h-mayorquin replied:
Yeah, I tried:

```python
 reference_contact = DynamicTableRegion(
        name="contacts",
        data=[2],
        description="The reference contact",
        table=contacts_table,
    )

```

And then using this as a entry in the row and it does not work.

This is probably a confusion of mine with respect to the schema but I wanted to take this opportunity to learn.

rly replied:
`contact` is a column that is a `DynamicTableRegion`. `ContactsTable.add_row(column=1)` adds a single entry `1` to that column (the "data" of the `DynamicTableRegion`). `add_row` takes the indices of the table. You cannot pass in a `DynamicTableRegion`. 

But you could initialize the `ContactsTable` with the `contacts` column with all the row values if you know them in advance:
```python
table = ContactsTable(columns=[
    DynamicTableRegion(
        name="contacts",
        data=[1, 2, 3, 4],
        description="The reference contacts",
        table=contacts_table,
    ), 
    # other columns
)
```

h-mayorquin replied:
Ah, great, so I can think on the row entry as the corresponding entry on a `DynamicTableRegion.data` . Thanks for educating me.

Relevant to this is the following. I remember vaguely that we discussed bipolar electrodes, for that case we would model it with `reference contact`. Is that correct? If that is the case, do you think it make sense to add some comments to the doc of `reference_contact` in that direction?

rly replied:
Good idea. I've updated the PR.

h-mayorquin replied:
Thanks.

h-mayorquin commented:
> Remaining TODOs for discussion.
> 
> * `ContactsTable.relative_position_in_mm.reference` ("Reference point for the relative position coordinates and information about the coordinate system used, e.g., which direction is positive in the x direction (first element), which direction is positive in the y direction (second element), etc.") is not required, but I wonder if it should be for maximal interpretability, especially if a tool builder wants to use these coordinates with the insertion position, angle, and depth of the probe to compute the stereotactic coordinates of each contact.

I am confused about this. The contacts table is just an non-spatial positioned version of geometry / shape of the probe. Why do we have an attribute for the `relative_position_in_mm`? Can we just have an attribute on the table itself indicating the origin? Would that not suffice?


> * Should `ChannelsTable.contact` (the reference to rows of the `ContactsTable`) be optional? In other words, should a `ContactsTable` be required to create an `ExtracellularSeries`. I think probably. We currently require `ElectrodeGroup` and `Device`. I'm just thinking about the number of lines of code this will add to the tutorials...

I think should be optional. @CodyCBakerPhD why do you think it should be required?  We are not sure if people will have the wiring map and we don't want them to make up information. The identity map is making up information.

h-mayorquin commented on spec/ndx-extracellular-channels.extensions.yaml:
```diff
@@ -0,0 +1,349 @@
+    doc: The brain area of the actual contact position, e.g., "CA1".
+    quantity: '?'
+  groups:
+  - name: probe_insertion
```
Can you guys @rly@ CodyCBakerPhD remind me why we decided to include the insertion in the channels table and not in the Probe? The probes are the things that are inserted so I think we had a good reason but I can't really recall. Do you?

rly replied:
I think the only reason might have been that because it contains position information, it is useful attached to the channels table where there is also position information. But I see your point that it is more a property of the `Probe`. Happy to make that change. (and tagging @CodyCBakerPhD correctly)

h-mayorquin replied:
Yes, I would prefer that. For the position information I am leaning to delegate that to your proposal in #3 of a `AnatomicalCoordinatesTable`.

rly committed:
Apply suggestions from code review

rly committed:
"reference_mode" -> "electrical_reference_mode"

rly committed:
Add "ground", rename "electrical_reference_description"

rly committed:
Remove "device_channel" from ContactsTable

rly committed:
Apply black

rly commented:
Thank you for the detailed review @h-mayorquin ! I have updated the PR based on your suggestions and discussion.

h-mayorquin reviewed (approved):
Ok, I read the round-trip tests and this looks OK to me. More than that, this is looking great. Thanks @rly for moving this forward. I learned a lot from your reviewing your code. Thanks for your patience.

There are things that I still don't like but that can be addressed later:
1) I think probe insertion should go into Probe as discussed. I think we agree on this and will also solve some concerns in https://github.com/catalystneuro/ndx-extracellular-channels/issues/3#issuecomment-2224585510 without too many changes.
2) Make the contacts optional. The wiring between channels and contacts is not obvious and in cases the users don't have it we should not force them to construct fake information. If the mapping is not available, then is not available and tha's it.
3) I don't really like having the estimated and confirmed positions as canonical attributes in the channels table. I think that the proposal on #3 of relegating this to an `AnatomicalCoordinatesTable` is better for the following reasons. First, separation of concerns between the description of the ecephys setup and experiments of localization. I think is nice to decouple those two things so they can evolve on their own. If the localization experiments improve we then only need the change to `AnatomicalCoordinatesTable` or link into a better extension without making parts of our schema obsolete. Second, in the case of SpikeGLX where two channels map to one contact we will end up with redundant information. Plus, even within the current schema I don't like the canonization of `estimated_{}` as a set of single columns. I think that `confirmed_{x}` is a great name but what if there is more than one estimate? Right now those are optional attributes and maybe something like this is needed urgently in which case is fine but I still wanted to point out what I think are flaws of this design.
3) Finally, I would still prefer the mapping between the channels table and device to be at the row level for the reasons described in https://github.com/catalystneuro/ndx-extracellular-channels/issues/3#issuecomment-2240622943.  Which in short are that the channels table is working as an abstraction of the DAQ (recording system plus headstage in most cases) and those devices usually map to more than one probe (like the DAQ of Openephys and Intan). In addition, this will make automatization of the conversion possible for some formats that don't have a way of telling if the data came from differentes probes (like Intan). Plus, mapping at the rows of the channel table to either device or rows of the `AnatomicalCoordinatesTable` would also solve the problems of duplication of information described on https://github.com/catalystneuro/ndx-extracellular-channels/issues/3#issue-2404114480 and is a more general workflow overall. I think is a little bit more clumsy but I think some of that can be attenuated with clever design at the API level.

What do you guys think?

h-mayorquin commented:
Also, I can take care of updating the diagram one we merged this.

h-mayorquin committed:
fix name

rly committed:
Update doc for "reference_contact" and angles

rly committed:
Move ProbeInsertion from ChannelsTable to Probe

rly committed:
Update mermaid diagram for probe_insertion move

rly committed:
Fix relative_position_in_mm should be in um

rly committed:
Remove ContactsTable.relative_position_in_um.reference

rly commented:
> 1. I think probe insertion should go into Probe as discussed. I think we agree on this and will also solve some concerns in [[Proposal] The return of ElectrodesTable #3 (comment)](https://github.com/catalystneuro/ndx-extracellular-channels/issues/3#issuecomment-2224585510) without too many changes.

I just moved `ProbeInsertion` to `Probe`. We can adjust it again in #3 if we want to.

> 2. Make the contacts optional. The wiring between channels and contacts is not obvious and in cases the users don't have it we should not force them to construct fake information. If the mapping is not available, then is not available and tha's it.

I'm not sure. `ChannelsTable` requires `Probe`, `Probe` requires `ProbeModel`, `ProbeModel` requires `ContactsTable`, and `ContactsTable` requires `relative_position_in_um`. So already the user must know the position of each contact on the probe. I think the user will also know the mapping of their data to contacts on a probe in most cases because that's usually needed to interpret the data. I get that's not always the case. For those who don't, we could propose that they specify an identity mapping and set the `relative_position_in_um` values to all `NaN` to indicate they don't know which contact is which.

People using tetrodes usually do not know the relative position of individual contacts. **Should `ContactsTable.relative_position_in_um` be optional?**

> 3. I don't really like having the estimated and confirmed positions as canonical attributes in the channels table. I think that the proposal on [[Proposal] The return of ElectrodesTable #3](https://github.com/catalystneuro/ndx-extracellular-channels/issues/3) of relegating this to an `AnatomicalCoordinatesTable` is better for the following reasons. First, separation of concerns between the description of the ecephys setup and experiments of localization. I think is nice to decouple those two things so they can evolve on their own. If the localization experiments improve we then only need the change to `AnatomicalCoordinatesTable` or link into a better extension without making parts of our schema obsolete. Second, in the case of SpikeGLX where two channels map to one contact we will end up with redundant information. Plus, even within the current schema I don't like the canonization of `estimated_{}` as a set of single columns. I think that `confirmed_{x}` is a great name but what if there is more than one estimate? Right now those are optional attributes and maybe something like this is needed urgently in which case is fine but I still wanted to point out what I think are flaws of this design.

Let's discuss this in #3 and move forward here.

> 4. Finally, I would still prefer the mapping between the channels table and device to be at the row level for the reasons described in [[Proposal] The return of ElectrodesTable #3 (comment)](https://github.com/catalystneuro/ndx-extracellular-channels/issues/3#issuecomment-2240622943).  Which in short are that the channels table is working as an abstraction of the DAQ (recording system plus headstage in most cases) and those devices usually map to more than one probe (like the DAQ of Openephys and Intan). In addition, this will make automatization of the conversion possible for some formats that don't have a way of telling if the data came from differentes probes (like Intan). Plus, mapping at the rows of the channel table to either device or rows of the `AnatomicalCoordinatesTable` would also solve the problems of duplication of information described on [[Proposal] The return of ElectrodesTable #3 (comment)](https://github.com/catalystneuro/ndx-extracellular-channels/issues/3#issue-2404114480) and is a more general workflow overall. I think is a little bit more clumsy but I think some of that can be attenuated with clever design at the API level.
> 
> What do you guys think?

Let's discuss this in #3 and move forward here.

Lastly, in reviewing `ContactsTable.relative_position_in_um.reference`, I think it is not necessary and so I have removed it. Let's discuss the use case for it in #4 with a clear example before adding it in.

Thank you so much @h-mayorquin and @CodyCBakerPhD for the thorough reviews! I just have one remaining question above. I suggest we discuss the rest of the topics as independent enhancements that we address after merging this already large PR.

CodyCBakerPhD commented:
@rly Thanks for pushing on this!
